tortoise/verifying_tortoise.go

```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
+	ctx context.Context,
+	baseBlockLayerID types.LayerID,
+	baseBlockOpinion Opinion, // candidate base block's opinion vector
+) ([]map[types.BlockID]struct{}, error) {
+	logger := t.logger.WithContext(ctx).WithFields(log.FieldNamed("base_block_layer_id", baseBlockLayerID))
+
 	// using maps prevents duplicates
 	againstDiff := make(map[types.BlockID]struct{})
 	forDiff := make(map[types.BlockID]struct{})
 	neutralDiff := make(map[types.BlockID]struct{})
 
-	// all genesis blocks get a for vote
-	if layerid == types.GetEffectiveGenesis() {
+	// we support all genesis blocks by default
+	if baseBlockLayerID == types.GetEffectiveGenesis() {
 		for _, i := range types.BlockIDs(mesh.GenesisLayer().Blocks()) {
 			forDiff[i] = struct{}{}
 		}
-		return []map[types.BlockID]struct{}{againstDiff, forDiff, neutralDiff}, nil
 	}
 
-	// TODO: maybe we should vote back hdist but drill down check the base blocks
-
-	// Add latest layers input vector results to the diff.
-	for i := layerid; !i.After(t.Last); i = i.Add(1) {
-		t.logger.With().Debug("checking input vector diffs", i)
+	// Add latest layers input vector results to the diff
+	// Note: a block may only be selected as a candidate base block if it's marked "good", and it may only be marked
+	// "good" if its own base block is marked "good" and all exceptions it contains agree with our local opinion.
+	// We only look for and store exceptions within the sliding window set of layers as an optimization, but a block
+	// can contain exceptions from any layer, back to genesis.
+	startLayer := t.LastEvicted.Add(1)
+	if startLayer.Before(types.GetEffectiveGenesis()) {
+		startLayer = types.GetEffectiveGenesis()
+	}
+	for layerID := startLayer; !layerID.After(t.Last); layerID = layerID.Add(1) {
+		logger := logger.WithFields(log.FieldNamed("diff_layer_id", layerID))
+		logger.Debug("checking input vector diffs")
 
-		layerBlockIds, err := t.bdp.LayerBlockIds(i)
+		layerBlockIds, err := t.bdp.LayerBlockIds(layerID)
 		if err != nil {
 			if err != leveldb.ErrClosed {
-				// todo: empty layer? maybe skip verify differently
-				t.logger.Panic("can't find old layer %v", i)
+				// this should not happen! we only look at layers up to the last processed layer, and we only process
+				// layers with valid block data.
+				logger.With().Error("no block ids for layer in database", log.Err(err))
 			}
 			return nil, err
 		}
 
-		// TODO: how to differentiate between missing Hare results, and Hare failed/gave up waiting?
-		layerInputVector, err := t.bdp.GetLayerInputVectorByID(i)
+		// helper function for adding diffs
+		addDiffs := func(bid types.BlockID, voteClass string, voteVec vec, diffMap map[types.BlockID]struct{}) {
+			if v, ok := baseBlockOpinion.BlockOpinions[bid]; !ok || simplifyVote(v) != voteVec {
+				logger.With().Debug("added vote diff",
+					log.FieldNamed("diff_block", bid),
+					log.String("diff_class", voteClass))
+				if layerID.Before(baseBlockLayerID) {
```

Comment:
should we allow this to happen? why not add the diff only if the layer is newer? it sounds to me it might be required for self healing.

Reply:
for a given base block candidate, it's essential that the diffs be added, that's how the verifying tortoise algorithm works. what we _could_ do instead is do a better job of picking base blocks, so as to minimize the length of the exception list. that's a to do item in #2402.